### PR TITLE
[Feature][Elastic search] Support multi-table source feature

### LINKED ISSUE #6730: [Feature][Elastic search] Support multi-table source feature
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

Elastic search Support multi-table source feature 

### Usage Scenario

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by FuYouJ**
### Purpose of this pull request

* Related issue https://github.com/apache/seatunnel/issues/6730
* Add Chinese document
* E2E allows spark to synchronize multiple tables,add more null value test case in e2e
* The configuration examples that included schema have been removed from the documentation, as it is strongly discouraged to use them.
image

### Check list

* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [x] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)
* [x] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java [hunk 1]
```diff
@@ -185,6 +186,28 @@ public void testElasticsearch(TestContainer container)
         Assertions.assertIterableEquals(mapTestDatasetForDSL(), sinkData);
     }
 
+    @TestTemplate
+    @DisabledOnContainer(
+            value = {},
+            type = {EngineType.SPARK, EngineType.FLINK},
+            disabledReason = "Currently SPARK/FLINK do not support multiple table read")
```
Comment (Carl-Zhou-CN) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java]: spark already supports multiple tables. You need to enable test case

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java [hunk 2]
```diff
@@ -304,17 +331,30 @@ private List<String> generateTestDataSet() throws JsonProcessingException {
...
+        List<String> source = new ArrayList<>(esFieldType.keySet());
+        return getDocsWithTransformDate(source, indexName, nullAllowedFields);
+    }
+
+    // The timestamp type in Elasticsearch is incompatible with that in Seatunnel,
+    // and we need to handle the conversion here.
+    private List<String> readSinkDataWithSchema(String index) throws InterruptedException {
         // wait for index refresh
-        Thread.sleep(2000);
+        Thread.sleep(5000);
```
Comment (Carl-Zhou-CN) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java:357]: Can we use a parameter here?

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java [hunk 3]
```diff
@@ -394,6 +434,12 @@ private List<String> getDocsWithTransformTimestamp(List<String> source, String i
     }
 
     private List<String> getDocsWithTransformDate(List<String> source, String index) {
+        return getDocsWithTransformDate(source, index, Collections.emptyList());
+    }
+
+    //
```
Comment (Carl-Zhou-CN) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java]: remove

File: seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/source/ElasticsearchSource.java [hunk 4]
```diff
@@ -58,28 +59,51 @@ public class ElasticsearchSource
...
-            source = config.get(SourceConfig.SOURCE);
-            arrayColumn = config.get(SourceConfig.ARRAY_COLUMN);
-            EsRestClient esRestClient = EsRestClient.createInstance(config);
+            source = readonlyConfig.get(SourceConfig.SOURCE);
+            arrayColumn = readonlyConfig.get(SourceConfig.ARRAY_COLUMN);
+            EsRestClient esRestClient = EsRestClient.createInstance(connectionConfig);
             Map<String, BasicTypeDefine<EsType>> esFieldType =
-                    esRestClient.getFieldTypeMapping(config.get(SourceConfig.INDEX), source);
+                    esRestClient.getFieldTypeMapping(index, source);
             esRestClient.close();
```
Comment (Carl-Zhou-CN) [inline seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/source/ElasticsearchSource.java]: Is there a risk that the connection is not closed

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java [hunk 5]
```diff
@@ -304,17 +331,30 @@ private List<String> generateTestDataSet() throws JsonProcessingException {
         return documents;
     }
 
-    private List<String> readSinkDataWithOutSchema() throws InterruptedException {
+    private List<String> readSinkDataWithOutSchema(String indexName) throws InterruptedException {
         Map<String, BasicTypeDefine<EsType>> esFieldType =
-                esRestClient.getFieldTypeMapping("st_index4", Lists.newArrayList());
-        Thread.sleep(2000);
+                esRestClient.getFieldTypeMapping(indexName, Lists.newArrayList());
+        Thread.sleep(5000);
```
Comment (Carl-Zhou-CN) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java]: Can unify a parameter

Comment (FuYouJ) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java]: Now all refresh waiting times use static variables.
```
private static final long INDEX_REFRESH_MILL_DELAY = 5000L;
```

File: seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/source/ElasticsearchSource.java (hunk 4 above)
Comment (FuYouJ) [reply seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/source/ElasticsearchSource.java]: I implemented Closeable for EsRestClient and extracted the code into a separate method to ensure that resources are closed even if an exception occurs
```
    private Map<String, BasicTypeDefine<EsType>> getFieldTypeMapping(
            String index, List<String> source) {
        // EsRestClient#getFieldTypeMapping may throw runtime exception
        // so here we use try-resources-finally to close the resource
        try (EsRestClient esRestClient = EsRestClient.createInstance(connectionConfig)) {
            return esRestClient.getFieldTypeMapping(index, source);
        }
    }
```

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java (hunk 3 above)
Comment (FuYouJ) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java]: done

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java (hunk 1 above)
Comment (FuYouJ) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/java/org/apache/seatunnel/e2e/connector/elasticsearch/ElasticsearchIT.java]: done

**Review by Carl-Zhou-CN — DISMISSED** [review]
LGTM

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/resources/elasticsearch/elasticsearch_multi_source_and_sink.conf [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+sink {
+  Elasticsearch {
+    hosts = ["https://elasticsearch:9200"]
+    username = "elastic"
+    password = "elasticsearch"
+    tls_verify_certificate = false
+    tls_verify_hostname = false
+
+    index = "multi_source_write_test_index"
+    index_type = "st"
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/resources/elasticsearch/elasticsearch_multi_source_and_sink.conf]: You should test this feature using multi-table writes

check read/write tables row & column

https://github.com/apache/seatunnel/pull/7052/files#diff-9c8de872b906856b7dccdbab2dbe4a0ca089be7ce5fd89097aeec8ba82540c27

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/resources/elasticsearch/elasticsearch_multi_source_and_sink.conf [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+           c_tinyint,
+           c_smallint,
+           c_bigint,
+           c_float,
+           c_double,
+           c_decimal,
+           c_bytes,
+           c_int,
+           c_date,
+           c_timestamp
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/resources/elasticsearch/elasticsearch_multi_source_and_sink.conf]: You should define indexes using different fields

e.g

```
index_1: x, y, z...
index_2: a,b,c...
```

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/resources/elasticsearch/elasticsearch_multi_source_and_sink.conf (hunk 6 above)
Comment (FuYouJ) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/resources/elasticsearch/elasticsearch_multi_source_and_sink.conf]: Based on your suggestion, I modified my e2e test.
Now it reads different fields from different indices and writes them into different target indices.
```hocon
env {
  parallelism = 1
  job.mode = "BATCH"
  #checkpoint.interval = 10000
}

source {
  Elasticsearch {
    hosts = ["https://elasticsearch:9200"]
    username = "elastic"
    password = "elasticsearch"
    tls_verify_certificate = false
    tls_verify_hostname = false
    index_list = [
       {
           index = "read_filter_index1"
           query = {"range": {"c_int": {"gte": 10, "lte": 20}}}
           source = [
           c_map,
           c_array,
           c_string,
           c_boolean,
           c_tinyint,
           c_smallint,
           c_bigint,
           c_float,
           c_double,
           c_decimal,
           c_bytes,
           c_int,
           c_date,
           c_timestamp,
           c_null
           ]
           array_column = {
           c_array = "array<tinyint>"
           }
       }
       {
           index = "read_filter_index2"
           query = {"range": {"c_int2": {"gte": 10, "lte": 20}}}
           source = [
           c_int2,
           c_null2,
           c_date2
           ]

       }

    ]

  }
}

transform {
}

sink {
  Elasticsearch {
    hosts = ["https://elasticsearch:9200"]
    username = "elastic"
    password = "elasticsearch"
    tls_verify_certificate = false
    tls_verify_hostname = false

    index = "${table_name}_copy"
    index_type = "st"
    "schema_save_mode"="CREATE_SCHEMA_WHEN_NOT_EXIST"
    "data_save_mode"="APPEND_DATA"
  }
}

```

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/resources/elasticsearch/elasticsearch_multi_source_and_sink.conf (hunk 7 above)
Comment (FuYouJ) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-elasticsearch-e2e/src/test/resources/elasticsearch/elasticsearch_multi_source_and_sink.conf]: Based on your suggestion, I modified my e2e test.
Now it reads different fields from different indices and writes them into different target indices.
```hocon
env {
  parallelism = 1
  job.mode = "BATCH"
  #checkpoint.interval = 10000
}

source {
  Elasticsearch {
    hosts = ["https://elasticsearch:9200"]
    username = "elastic"
    password = "elasticsearch"
    tls_verify_certificate = false
    tls_verify_hostname = false
    index_list = [
       {
           index = "read_filter_index1"
           query = {"range": {"c_int": {"gte": 10, "lte": 20}}}
           source = [
           c_map,
           c_array,
           c_string,
           c_boolean,
           c_tinyint,
           c_smallint,
           c_bigint,
           c_float,
           c_double,
           c_decimal,
           c_bytes,
           c_int,
           c_date,
           c_timestamp,
           c_null
           ]
           array_column = {
           c_array = "array<tinyint>"
           }
       }
       {
           index = "read_filter_index2"
           query = {"range": {"c_int2": {"gte": 10, "lte": 20}}}
           source = [
           c_int2,
           c_null2,
           c_date2
           ]

       }

    ]

  }
}

transform {
}

sink {
  Elasticsearch {
    hosts = ["https://elasticsearch:9200"]
    username = "elastic"
    password = "elasticsearch"
    tls_verify_certificate = false
    tls_verify_hostname = false

    index = "${table_name}_copy"
    index_type = "st"
    "schema_save_mode"="CREATE_SCHEMA_WHEN_NOT_EXIST"
    "data_save_mode"="APPEND_DATA"
  }
}

```

File: seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/source/ElasticsearchSourceFactory.java [hunk 8]
```diff
@@ -55,7 +52,7 @@ public String factoryIdentifier() {
     @Override
     public OptionRule optionRule() {
         return OptionRule.builder()
-                .required(HOSTS, INDEX)
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/source/ElasticsearchSourceFactory.java:58]: Is it compatible with old versions?

File: docs/en/connector-v2/source/Elasticsearch.md [hunk 9]
```diff
@@ -19,24 +19,25 @@ support version >= 2.x and <= 8.x.
 
 ## Options
 
-|          name           |  type   | required |   default value   |
```
Comment (hailin0) [inline docs/en/connector-v2/source/Elasticsearch.md:22]: Why is the `index_list` not added?

**Comment by hailin0** [comment]
What will happen if I configure index & index_list connector at the same time

File: docs/en/connector-v2/source/Elasticsearch.md (hunk 9 above)
Comment (FuYouJ) [reply docs/en/connector-v2/source/Elasticsearch.md:22]: image
after scroll_size

File: seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/source/ElasticsearchSourceFactory.java (hunk 8 above)
Comment (FuYouJ) [reply seatunnel-connectors-v2/connector-elasticsearch/src/main/java/org/apache/seatunnel/connectors/seatunnel/elasticsearch/source/ElasticsearchSourceFactory.java:58]: Fully compatible with previous configurations, in the ElasticsearchSource code, the first if step is to check for the existence of index_ist. If it exists, it should be parsed synchronously as multiple tables, and if it does not exist, it should be parsed synchronously as a single table
```
    public ElasticsearchSource(ReadonlyConfig config) {
        this.connectionConfig = config;
        boolean multiSource = config.getOptional(SourceConfig.INDEX_LIST).isPresent();
        boolean singleSource = config.getOptional(SourceConfig.INDEX).isPresent();
        if (multiSource && singleSource) {
            log.warn(
                    "Elasticsearch Source config warn: when both 'index' and 'index_list' are present in the configuration, only the 'index_list' configuration will take effect");
        }
        if (!multiSource && !singleSource) {
            throw new ElasticsearchConnectorException(
                    ElasticsearchConnectorErrorCode.SOURCE_CONFIG_ERROR_01,
                    ElasticsearchConnectorErrorCode.SOURCE_CONFIG_ERROR_01.getDescription());
        }
        if (multiSource) {
            this.sourceConfigList = createMultiSource(config);
        } else {
            this.sourceConfigList = Collections.singletonList(parseOneIndexQueryConfig(config));
        }
    }
```

File: docs/en/connector-v2/source/Elasticsearch.md (hunk 9 above)
Comment (hailin0) [reply docs/en/connector-v2/source/Elasticsearch.md:22]: ok

**Comment by FuYouJ** [comment]
> What will happen if I configure index & index_list connector at the same time

The program will print a warning log to tell the user the configured processing priority。
```
f (multiSource && singleSource) {
            log.warn(
                    "Elasticsearch Source config warn: when both 'index' and 'index_list' are present in the configuration, only the 'index_list' configuration will take effect");
        }
```

**Review by hailin0 — APPROVED** [review]
LGTM. Thanks @FuYouJ 

Waiting for ci passed

**Review by Carl-Zhou-CN — APPROVED** [review]
+1

**Comment by Carl-Zhou-CN** [comment]
Thank you for your contribution @FuYouJ

**Pull request merged**